activesupport/lib/active_support/inspect.rb

```diff
@@ -0,0 +1,378 @@
+  # variables may be accessed by their "@"-prefixed name.
+  #
+  # Attributes can also be specified as a hash, in which case the key is
+  # the label to be displayed, and the value is the attribute to be
+  # inspected (or a proc that will be called to obtain the value).
+  #
+  # Before the attributes appear in the output, space is reserved for an
+  # optional object +label+, which is intended to be a short value
+  # identifying the object instance.
+  #
+  # To customize the label, pass a block that returns a string, or set
+  # the +label:+ parameter to a symbol to call that method on the
+  # instance being inspected.
+  #
+  # +label+ may produce an array, in which case the elements will be
+  # rendered in order, with each element separated by whitespace.
+  # Strings are inserted in the output unmodified; this provides more
+  # control over the output, but means #inspect should be used for
+  # displaying unknown values to retain proper formatting. Any element
+  # that is not an array or a string will be recursively pretty-printed.
+  #
+  # By default the built-in object address (e.g. 0x0123456789abcdef),
+  # which appears in the standard Kernel#inspect output, is not
+  # included. To add it, pass +id: true+. This can be useful when
+  # objects are otherwise indistinguishable, but makes for more
+  # cluttered output. The expectation is that most classes defining
+  # custom inspection will provide sufficient detail in the label and/or
+  # attributes to identify the instance.
+  #
+  #
+  # When this module is included, if the target class/module does not
+  # have a specific #to_s method, an additional ToString module will
+  # automatically be included, which will provide a #to_s method that
+  # produces a subset of the #inspect output. (Specifically, the class
+  # name and object id / label, but not the attributes.)
+  #
+  #
+  # ==== Example
+  #
+  #  class User
+  #    def initialize(id, username, name, interests, password_digest)
+  #      @id = id
+  #      @username = username
+  #      @name = name
+  #      @interests = interests
+  #      @password_digest = password_digest
+  #    end
+  #  end
+  #
+  #  user = User.new(77, "jdoe", "John Doe", ["sports", "music"],
+  #                  "61bc77ec652ed6ace7a8eb44...")
+  #
+  #  # Default Ruby behavior:
+  #
+  #  user.inspect #=> "#<User:0x00007f9b8f8b9840 @id=77, @username=\"jdoe\", @name=\"John Doe\", @interests=[\"sports\", \"music\"], @password_digest=\"61bc77ec652ed6ace7a8eb44...\">"
+  #  user.to_s #=> "#<User:0x00007f9b8f8b9840>"
+  #
+  #  pp user #=>
+  #  > #<User:0x0000000113695bb8
+  #  >  @id=77,
+  #  >  @interests=["sports", "music"],
+  #  >  @name="John Doe",
+  #  >  @password_digest="61bc77ec652ed6ace7a8eb44...",
+  #  >  @username="jdoe">
+  #
+  #  # After defining a custom Inspect for the User class:
+  #
+  #  class User
+  #    include ActiveSupport::Inspect(:@name, { interests: -> { @interests.size } }) { [@id, @username.to_s] }
```

Comment:
What are your thoughts on making the API entirely block-based?  e.g.:

```ruby
include ActiveSupport::Inspect { [[@id, @username.to_s], :@name, interests: @interests.size] }
# => "#<User 77 \"jdoe\" @name=\"John Doe\" interests=2>"
```

Elements that are symbols or hashes would be resolved / rendered in the same way.  Elements that are strings would be rendered as-is.  Elements that are arrays would be `.map(&:inspect).join(" ")`'d.  And, if the first element is a string or array, it would be treated as the label.

And these could be interleaved:

```ruby
include ActiveSupport::Inspect { [[@id], { interests: @interests.size }, :@username, "(#{@name})"] }
# => "#<User 77 interests=2 @username=\"jdoe\" (John Doe)>"
```

This would also prevent conflicts with any kwargs, e.g. `ActiveSupport::Inspect(id: true)`.